## Find the Bug

Let's look at a piece of code that demonstrates additional interface mechanics that are nice to know.

```go
package main

import "log"

// customError is just an empty struct
type customError struct{}

// Error implements the error interface
func (c *customError) Error() string {
	return "Find the bug."
}

// fail returns nil values for both return types
func fail() ([]byte, *customError) {
	return nil, nil
}

func main() {
	var err error
	if _, err = fail(); err != nil {
		log.Fatal("Why did this fail?")
	}

	log.Println("No Error!")
}
```

When `fail()` returns a nil pointer for a specific type (`*customError`), it is not equivalent to a nil `error` interface. This is because an interface in Go is nil only if both its type and value are nil. In the case of `fail()`, the interface `err` receives a nil value of type `*customError`, making the interface itself non-nil. This is why the check `err != nil` unexpectedly evaluates to true.

```go
package main

import "log"

// fail returns nil values for both return types
func fail() ([]byte, error) {
	return nil, nil
}

func main() {
	var err error
	if _, err = fail(); err != nil {
		log.Fatal("Why did this fail?")
	}

	log.Println("No Error!")
}
```

